zsh_completions_v2.go

```diff
@@ -0,0 +1,145 @@
+        # We are dealing with a flag with an =
+        flagPrefix=${BASH_REMATCH}
+    fi
+
+    # Prepare the command to obtain completions
+    requestComp="${words[1]} %[2]s ${words[2,-1]}"
+    if [ "${lastChar}" = "" ]; then
+        # If the last parameter is complete (there is a space following it)
+        # We add an extra empty parameter so we can indicate this to the go completion code.
+        __%[1]s_debug "Adding extra empty parameter"
+        requestComp="${requestComp} \"\""
+    fi
+
+    __%[1]s_debug "About to call: eval ${requestComp}"
+
+    # Use eval to handle any environment variables and such
+    out=$(eval ${requestComp} 2>/dev/null)
+    __%[1]s_debug "completion output: ${out}"
+
+    # Extract the directive integer following a : from the last line
+    if [ "${out[-2]}" = : ]; then
+        directive=${out[-1]}
+        # Remove the directive (that means the last 3 chars as we include the : and the newline)
+        out=${out[1,-4]}
+    else
+        # There is no directive specified.  Leave $out as is.
+        __%[1]s_debug "No directive found.  Setting do default"
+        directive=0
+    fi
+
+    __%[1]s_debug "directive: ${directive}"
+    __%[1]s_debug "completions: ${out}"
+    __%[1]s_debug "flagPrefix: ${flagPrefix}"
+
+    if [ $((directive & %[3]d)) -ne 0 ]; then
+        __%[1]s_debug "Completion received error. Ignoring completions."
+        return
+    fi
+
+    compCount=0
+    while IFS='\n' read -r comp; do
+        if [ -n "$comp" ]; then
+            # If requested, completions are returned with a description.
+            # The description is preceded by a TAB character.
+            # For zsh's _describe, we need to use a : instead of a TAB.
+            # We first need to escape any : as part of the completion itself.
+            comp=${comp//:/\\:}
+
+            local tab=$(printf '\t')
+            comp=${comp//$tab/:}
+
+            ((compCount++))
+            if [ -n "$flagPrefix" ]; then
+                __%[1]s_debug "Adding completion: ${flagPrefix}${comp}"
+                completions+="${flagPrefix}${comp}"
+            else
+                __%[1]s_debug "Adding completion: ${comp}"
+                completions+=${comp}
+            fi
+            lastComp=$comp
+        fi
+    done < <(printf "%%s\n" "${out[@]}")
+
+    if [ ${compCount} -eq 0 ]; then
+        if [ $((directive & %[5]d)) -ne 0 ]; then
+            __%[1]s_debug "deactivating file completion"
+        else
+            # Perform file completion
+            __%[1]s_debug "activating file completion"
+            _arguments '*:filename:_files'
+        fi
+    elif [ $((directive & %[4]d)) -ne 0 ] && [ ${compCount} -eq 1 ]; then
+        __%[1]s_debug "Activating nospace."
+        # We can use compadd here as there is no description when
+        # there is only one completion.
+        compadd -S '' "${lastComp}"
+    else
+        _describe "completions" completions
+    fi
+}
+
+compdef _%[1]s %[1]s
```

Comment:
This line allows to source the zsh completion script as is currently done for `helm` and `kubectl`.  I like the idea of providing that option.

However, I need to point out that @jharshman disagreed with this in https://github.com/spf13/cobra/pull/887#issuecomment-536621700, so he may prefer to have this removed, once we get to review time.

Reply:
@marckhouzam yeah I still maintain my previous opinion about this.

Reply:
@jharshman is right.  I found this article to be a good explanation of why sourcing a completion script should not be recommended: https://medium.com/@jzelinskie/please-dont-ship-binaries-with-shell-completion-as-commands-a8b1bcb8a0d0

For projects that need to keep backwards-compatibility, they can insert that one extra line themselves.  I will need to do that for Helm.

I will remove this line in my next update.